// @docknetwork/crypto-wasm-ts/lib/anonymous-credentials/types-and-consts.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SAVER_ENCRYPTION_GENS_BYTES = exports.SAVER_ENCRYPTION_GENS_LABEL = exports.ACCUMULATOR_PROVING_KEY_LABEL_BYTES = exports.ACCUMULATOR_PROVING_KEY_LABEL = exports.ACCUMULATOR_PARAMS_LABEL_BYTES = exports.ACCUMULATOR_PARAMS_LABEL = exports.SIGNATURE_PARAMS_LABEL_BYTES = exports.SIGNATURE_PARAMS_LABEL = exports.CRED_PROOF_TYPE = exports.NON_MEM_CHECK_STR = exports.MEM_CHECK_STR = exports.REV_ID_STR = exports.REV_CHECK_STR = exports.ID_STR = exports.STATUS_TYPE_STR = exports.TYPE_STR = exports.STATUS_STR = exports.SUBJECT_STR = exports.SCHEMA_TYPE_STR = exports.SCHEMA_STR = exports.CRYPTO_VERSION_STR = exports.VERSION_STR = void 0;
exports.VERSION_STR = 'version';
exports.CRYPTO_VERSION_STR = 'cryptoVersion';
exports.SCHEMA_STR = 'credentialSchema';
exports.SCHEMA_TYPE_STR = 'JsonSchemaValidator2018';
exports.SUBJECT_STR = 'credentialSubject';
exports.STATUS_STR = 'credentialStatus';
exports.TYPE_STR = 'type';
exports.STATUS_TYPE_STR = 'DockVBAccumulator2022';
exports.ID_STR = 'id';
exports.REV_CHECK_STR = 'revocationCheck';
exports.REV_ID_STR = 'revocationId';
exports.MEM_CHECK_STR = 'membership';
exports.NON_MEM_CHECK_STR = 'non-membership';
exports.CRED_PROOF_TYPE = 'Bls12381BBS+SignatureDock2022';
const te = new TextEncoder();
// Label used for generating signature parameters
exports.SIGNATURE_PARAMS_LABEL = 'DockBBS+Signature2022';
exports.SIGNATURE_PARAMS_LABEL_BYTES = te.encode(exports.SIGNATURE_PARAMS_LABEL);
// Label used for generating accumulator parameters
exports.ACCUMULATOR_PARAMS_LABEL = 'DockVBAccumulator2022';
exports.ACCUMULATOR_PARAMS_LABEL_BYTES = te.encode(exports.ACCUMULATOR_PARAMS_LABEL);
// Label used for generating accumulator proving key
exports.ACCUMULATOR_PROVING_KEY_LABEL = 'DockVBAccumulatorProvingKey2022';
exports.ACCUMULATOR_PROVING_KEY_LABEL_BYTES = te.encode(exports.ACCUMULATOR_PROVING_KEY_LABEL);
// Label used for generating SAVER encryption generators
exports.SAVER_ENCRYPTION_GENS_LABEL = 'DockSAVEREncryptionGens2022';
exports.SAVER_ENCRYPTION_GENS_BYTES = te.encode(exports.SAVER_ENCRYPTION_GENS_LABEL);
